BeagleBone_black/src/sensors/vl6180.cpp

```diff
@@ -0,0 +1,211 @@
+constexpr uint16_t FIRMWARE__RESULT_SCALER               = 0x0120;
+constexpr uint16_t I2C_SLAVE__DEVICE_ADDRESS             = 0x0212;
+constexpr uint16_t INTERLEAVED_MODE__ENABLE              = 0x02A3;
+
+namespace hyped {
+namespace sensors {
+
+Vl6180::Vl6180(uint8_t id, Logger& log)
+  :Thread(id, log)
+{
+  log_.INFO("VL6180", "Creating a sensor with id: %d", id);
+}
+
+void Vl6180:: turnOn()
+{
+  // return if already on
+  if(this->on_) {
+    log_.DBG("VL6180", "Sensor is already on\n");
+    return;
+  }
+
+  // Wait incase the sensor has just been turned off
+  // TODO need to check wait time
+
+  // Turn on pins TODO pin write to turn sensor on
+
+  // Wait for 1.5ms (Data sheet says 1.4ms)
+  this->waitDeviceBooted();
+
+  // Initialise the sensor / register tuning
+  // Taken from ST Microelectronics API
+  this->writeByte(0x0207, 0x01);
+  this->writeByte(0x0208, 0x01);
+  this->writeByte(0x0096, 0x00);
+  this->writeByte(0x0097, 0xfd);
+  this->writeByte(0x00e3, 0x00);
+  this->writeByte(0x00e4, 0x04);
+  this->writeByte(0x00e5, 0x02);
+  this->writeByte(0x00e6, 0x01);
+  this->writeByte(0x00e7, 0x03);
+  this->writeByte(0x00f5, 0x02);
+  this->writeByte(0x00d9, 0x05);
+  this->writeByte(0x00db, 0xce);
+  this->writeByte(0x00dc, 0x03);
+  this->writeByte(0x00dd, 0xf8);
+  this->writeByte(0x009f, 0x00);
+  this->writeByte(0x00a3, 0x3c);
+  this->writeByte(0x00b7, 0x00);
+  this->writeByte(0x00bb, 0x3c);
+  this->writeByte(0x00b2, 0x09);
+  this->writeByte(0x00ca, 0x09);
+  this->writeByte(0x0198, 0x01);
+  this->writeByte(0x01b0, 0x17);
+  this->writeByte(0x01ad, 0x00);
+  this->writeByte(0x00ff, 0x05);
+  this->writeByte(0x0100, 0x05);
+  this->writeByte(0x0199, 0x05);
+  this->writeByte(0x01a6, 0x1b);
+  this->writeByte(0x01ac, 0x3e);
+  this->writeByte(0x01a7, 0x1f);
+  this->writeByte(0x0030, 0x00);
+
+  // Enables polling for New Sample ready when measurement completes
+  this->writeByte( 0x0011, 0x10);
+
+  // Set the averaging sample period (datasheet recommends 48)
+  this->writeByte(0x010a, 0x30);
+
+  // Perform a single recalibration
+  this->writeByte(SYSRANGE__VHV_RECALIBRATE, 0x01);
+
+  this->on_ = true;
+  log_.DBG("VL6180", "Sensor is on\n");
+}
+
+void Vl6180::turnOff()
+{
+  // TODO do pin write to turn off vl6180
+  this->on_ = false;
+  log_.DBG("VL6180", "Sensor is now off\n");
+}
+
+double Vl6180::getDistance()
+{
+  return 0;
+}
+
+bool Vl6180::waitDeviceBooted()
+{
+  // Will hold the return value of the register SYSTEM__FRESH_OUT_OF_RESET
+  uint8_t fresh_out_of_reset;
+  int status;
+
+  do
+  {
+    status = this->readByte(SYSTEM__FRESH_OUT_OF_RESET, &fresh_out_of_reset);
+  }
+  while(fresh_out_of_reset != 1 && status == 0);
+
+  return true;
+}
+
+bool Vl6180::rangeWaitDeviceReady()
+{
+  return false;
+}
+
+int Vl6180::readByte(uint16_t reg_add, uint8_t *data)
+{
+  char buffer[2];
+  buffer[0] = reg_add >> 8;
+  buffer[1] = reg_add & 0xFF;
+  char recv_buffer[1];
+
```

Comment:
I am adding __attribute__((unused)) to `buffer` and `recv_buffer` as the compiler is throwing the warning [-Wunused-but-set-variable]. Please remove this when the code to use it is added.